eth/stagedsync/stage_bor_heimdall.go

```diff
@@ -142,33 +142,29 @@ func BorHeimdallForward(
 		}
 	}
 
-	if mine {
-		minedHeader := cfg.miningState.MiningBlock.Header
-
-		if minedHeadNumber := minedHeader.Number.Uint64(); minedHeadNumber > headNumber {
-			// Whitelist service is called to check if the bor chain is
-			// on the cannonical chain according to milestones
-			if service != nil {
-				if !service.IsValidChain(minedHeadNumber, []*types.Header{minedHeader}) {
-					logger.Debug("[BorHeimdall] Verification failed for mined header", "hash", minedHeader.Hash(), "height", minedHeadNumber, "err", err)
-					dataflow.HeaderDownloadStates.AddChange(minedHeadNumber, dataflow.HeaderInvalidated)
-					s.state.UnwindTo(minedHeadNumber-1, ForkReset(minedHeader.Hash()))
-					return fmt.Errorf("mining on a wrong fork %d:%x", minedHeadNumber, minedHeader.Hash())
-				}
-			}
-		} else {
-			return fmt.Errorf("attempting to mine %d, which is behind current head: %d", minedHeadNumber, headNumber)
-		}
-	}
+	// if mine {
```

Comment:
Need some discussion around this (more details in the PR description - part 3).

Reply:
We can safely skip this checks as the block anyways goes through this stage (with `mine = false`) as if it's syncing.

Reply:
@manav2401 bor heimdall stage is called with `mine=true` in the mining stage loop. There are 2 separate loops - syncing and mining. My understanding is that when we are in the `mine` mode we need to get the header of the mined block from the mining state using `cfg.miningState.MiningBlock.Header`. However now that the mine block stage has been moved after the bor heimdall stage that means the mined header will not be available? We need the header in order to fetch and persist the state sync events I believe. This makes me think that we would need 2 stages - 1 before the mine block stage to fetch and persist spans and 1 after the mine block to fetch the state sync events. What are your thoughts & is my understanding so far correct?

Reply:
(note I believe currently we are not correctly handling the state sync events for the `mine=true` case as the mined header is not correctly passed to the `fetchAndWriteBorEvents` function)

Reply:
thinking about it more maybe splitting is not necessary since in iteration N the bor heimdall stage will process state sync events for the block created in iteration N-1 since the header for that block will be available

Reply:
Hey @taratorio, you're correct that we need header for processing certain things in the bor heimdall stage. I was initially of the view that we'll anyways pass through the same stage when syncing (because when we mine block N, the data like header, bodies, exec results are cached and then it's again sent to all the sync stages where most of them just directly use the cached values and persists the data in DB and perform the remaining operations). That time we do have a header available and that should be sufficient to perform all the validations. 

Re. you questions, I don't think we need the whole header looking at the `fetchAndWriteBorEvents` function. But, I am a bit confused because I didn't check that part of the code and looks like writing these things (not snapshot, but writing in contracts) should happen via bor consensus and not an external bor heimdall stage. At least, that's how it's done in bor. In this function, we could definitely persist the data (i.e. the state sync events) in snapshot and then bor consensus should be able to utilise it when required. Wdyt?